proposals/p2173.md

```diff
@@ -265,20 +276,422 @@ Rewrites aren't performed on the left-hand side of such an `=`, so
+they are different types that have an isomorphic set of values, somewhat like
+`i32` and `u32`. An `=` constraint is not merely learning a new fact about a
+type, it is requesting different behavior.
+
+### Combining constraints with `&`
+
+Suppose we have `X = C where .R = A` and `Y = C where .R = B`. What should
+`C & X` produce? What should `X & Y` produce?
+
+We could perhaps say that `X & Y` results in a constraint where the type of `R`
+has the union of the interface of `A` and the interface of `B`, and that `C & X`
+similarly results in a constraint where the type of `R` has the union of the
+interface of `A` and the interface originally specified by `C`. However, this
+proposal suggests a simpler rule:
+
+-   Combining two rewrite rules with different rewrite targets results in a
+    constraint where the associated constant is ambiguous. Given `T:! X & Y`,
+    the type expression `T.R` is ambiguous between a rewrite to `A` and a
+    rewrite to `B`. But given `T:! X & X`, `T.R` is unambiguously rewritten to
+    `A`.
+-   Combining a constraint with a rewrite rule with a constraint with no rewrite
+    rule preserves the rewrite rule. For example, supposing that
+    `interface Container` extends `interface Iterable`, and `Iterable` has an
+    associated constant `Element`, the constraint
+    `Container & (Iterable where .Element = i32)` is the same as the constraint
+    `(Container & Iterable) where .Element = i32` which is the same as the
+    constraint `Container where .Element = i32`.
+
+### Combining constraints with `and`
+
+It's possible for one `=` constraint in a `where` to refer to another. When this
+happens, the constraint `C where A and B` is interpreted as
+`(C where A) where B`, so rewrites in `A` are applied to names in `B`, but
+rewrites in `B` are not applied to names in `A`:
+
+```
+interface C {
+  let T:! Type;
+  let U:! Type;
+}
+// ✅ Same as `(C where .T = i32) where .U = .T`, which is
+// the same as `(C where .T = i32) where .U = i32`, which is
+// the same as `C where .T = i32 and .U = i32`.
+fn F[A:! C where .T = i32 and .U = .T]() {}
+```
+
+If a constraint is formed where a rewrite would apply to the right-hand side of
+another rewrite in the same constraint, but is not applied because it was not in
+scope, the program is invalid.
+
+```
+// ❌ Rewritten member `.U` is mentioned in a context
+// where its rewrite is not in scope.
+fn BadG[A:! C where .T = .U and .U = i32]() {}
+// ❌ Rewritten member `.U` is mentioned in a context
+// where its rewrite is not in scope.
+fn BadH[A:! C where .T = .U and .U = .T]() {}
+// ❌ Rewritten member `.T` is mentioned in a context
+// where its rewrite is not in scope.
+fn BadI[A:! (C where .T = i32) & (C where .U = .T)]() {}
+```
+
+In more detail, what happens here is:
+
+1.  The constraint is fully elaborated. `.X` is replaced with `.Self.X`. For
+    example, `C where .T = .U and .U = i32` is rewritten to
+    `C where .T = .Self.U and .U = i32`.
```

Comment:
Weird this only happens to the right-hand side of the `=`

Reply:
The left-hand side is not an expression, it's the name of an associated constant, so this rewrite isn't possible there. But in any case, I felt this description didn't add anything under the new constraint resolution rules, so I removed it.